proposals/sycl_modules.md

```diff
@@ -693,66 +723,184 @@ value to the specialization constant.
-to hide the value as a parameter, keeping the lambda-capture the same
-irrespectively of the support of the feature.
+For the module associated with the host device, and possibly any device with non native spec constant support, the value is carried by the `spec_constant` object it-self.
+For modules natively supporting specialization constant, implementations must guaranty the values returned by the usage of the spec_constant materialized into a constant.
+
+
+### Setting the value of spec constant
+
+To modify the value of a spec constant, the following API is available.
 
-The interface to set values to the specialization constant
-in the module is backend specific:
+Given a SYCL module, a spec constant can be set by the `sycl::set_spec_constant` function.
 
 ```cpp
 namespace sycl {
-namespace opencl {
-  template<typename ID, typename T>
-  spec_constant<ID, T> set_spec_constant(module<module_status::executable> syclModule, T cst) const;
-}  // namespace opencl
+
+template<class T, spec_id<T>& s>
+void set_spec_constant(module<module_status::input> syclModule, T);
+
+}  // namespace sycl
+```
+If C++17 or later is enabled, then the following overload is available
+```cpp
+namespace sycl {
+
+template<auto& s>
+void set_spec_constant(module<module_status::input> syclModule,
+                       typename std::remove_reference_t<decltype(s)>::type);
+
+}  // namespace sycl
+```
+
+It is valid for a user to set several times a spec constant with the same `spec_id`,
+only the last version of the value will be consider when building the module.
+Calling this function is not allowed during the submission of a task.
+
+### Getting the value of spec constant in kernel
+
+To retrive the value of a spec constant in a SYCL kernel context, the following API is available.
+
+```cpp
+namespace sycl {
+
+template<class T, spec_id<T>& s>
+T get_spec_constant(const spec_constant<T, s>&);
+
 }  // namespace sycl
 ```
+If C++17 or later is enabled, then the following overload is available
+```cpp
+namespace sycl {
+
+template<auto& s>
+typename std::remove_reference_t<decltype(s)>::type
+get_spec_constant(const spec_constant<typename std::remove_reference_t<decltype(s)>::type, s>&);
+
+}  // namespace sycl
+```
+
+This call is only valid whitin a SYCL kernel context and the `spec_constant` value must have been
+created by the handler that invoked the kernel from which the `sycl::get_spec_constant` is called.
+
+If the module has native support for specialization constant,
+the returned value becomes a constant value available for the runtime compiler to use for optimisations.
+
+**Note:** the value may become a constant (known numerical value), although this may only happen at runtime.
+From the SYCL compiler's perspective (C++ compiler) the value is understood to be a runtime value and
+for this reason it is not possible to use such value in a `constexpr`.
+
+### Retrieving a spec constant placeholder
```

Comment:
What happens for a program such as:
```c++
spec_id<int> id;
queue.submit([&](handler h) {
    spec_constant<int, id> c1 = h.set_spec_constant<int, id>(1);
    spec_constant<int, id> c2 = h.set_spec_constant<int, id>(2);
    h.single_task([=]{ 
        int v = c1.get_spec_constant();
    });
});
```
Is this a valid program? What's the value of `v`? Is it 1 if `native_spec_constant()` is false and 2 if `native_spec_constant()` is true?

Reply:
> Is this a valid program?

I think this should be banned, this look very odd and may induce unnecessary complexity on runtimes (whether there is native support or not). I'll clarify this.

> Is it 1 if native_spec_constant() is false and 2 if native_spec_constant() is true?

In all cases the result of `native_spec_constant()` does not affect the behavior.

Reply:
> What do you mean with ban?

Runtime exception.

> UB or require compiler error? I think the first isn't reasonable and the 2nd isn't possible.

i agree with you, I don't see why we would have an UB here and a compiler error is not always possible for this.